nomad/structs/structs.go

```diff
@@ -5568,24 +5568,51 @@ func (t *Task) Warnings() error {
 	return mErr.ErrorOrNil()
 }
 
+// TaskKind identifies the special kinds of tasks using the following format:
+// '<kind_name>(:<identifier>)`. The TaskKind can optionally include an identifier that
+// is opague to the Task. This identier can be used to relate the task to some
+// other entity based on the kind.
+//
+// For example, a task may have the TaskKind of `connect-proxy:service` where
+// 'connect-proxy' is the kind name and 'service' is the identifier that relates the
+// task to the service name of which it is a connect proxy for.
 type TaskKind string
 
-const connectPrefix = "connect-proxy:"
+// Name returns the kind name portion of the TaskKind
+func (k TaskKind) Name() string {
+	return strings.Split(string(k), ":")[0]
+}
 
-func (k TaskKind) IsConnect() bool {
-	return strings.HasPrefix(string(k), connectPrefix) && len(k) > len(connectPrefix)
+// Value returns the identifier of the TaskKind or an empty string if it doesn't
+// include one.
+func (k TaskKind) Value() string {
+	if s := strings.Split(string(k), ":"); len(s) > 1 {
+		return strings.Join(s[1:], ":")
+	}
+	return ""
+}
+
+func (k TaskKind) IsConnectProxy() (bool, TaskKindConnectProxy) {
+	if strings.HasPrefix(string(k), ConnectProxyPrefix+":") && len(k) > len(ConnectProxyPrefix)+1 {
+		return true, TaskKindConnectProxy{k}
+	}
+
+	return false, TaskKindConnectProxy{}
+}
+
+const ConnectProxyPrefix = "connect-proxy"
+
+// TaskKindProxyConnect is a TaskKind of 'connect-proxy'
+type TaskKindConnectProxy struct {
```

Comment:
Since `Value()` is on TaskKind I have a slight preference for dropping this type and just making `validateProxyService(TaskKind, []*Services)` a plain old func. We're getting a lot of little types that encapsulate very little, and that might be more confusing than its worth.

No big deal.